components/encryption/src/io.rs

```diff
@@ -422,25 +553,69 @@ mod tests {
         for method in methods {
             let key = generate_data_key(method);
             let readable_text = std::io::Cursor::new(plaintext.clone());
-            let (encrypter, iv) = EncrypterReader::new(readable_text, method, &key).unwrap();
+            let iv = Iv::new_ctr();
+            let encrypter = EncrypterReader::new(readable_text, method, &key, iv).unwrap();
             let mut decrypter = DecrypterReader::new(encrypter, method, &key, iv).unwrap();
-            let mut piece = vec![0; 10240];
+            let mut read = vec![0; 10240];
             for offset in offsets {
                 for size in sizes {
                     assert_eq!(
                         decrypter.seek(SeekFrom::Start(offset as u64)).unwrap(),
                         offset as u64
                     );
                     let actual_size = std::cmp::min(plaintext.len().saturating_sub(offset), size);
-                    assert_eq!(decrypter.read(&mut piece[..size]).unwrap(), actual_size);
+                    assert_eq!(decrypter.read(&mut read[..size]).unwrap(), actual_size);
                     if actual_size > 0 {
-                        assert_eq!(
-                            piece[..actual_size],
-                            plaintext[offset..offset + actual_size]
-                        );
+                        assert_eq!(read[..actual_size], plaintext[offset..offset + actual_size]);
                     }
                 }
             }
         }
     }
+
+    #[test]
+    fn test_encrypt_then_decrypt_write_plaintext() {
+        let methods = [
+            EncryptionMethod::Plaintext,
+            EncryptionMethod::Aes128Ctr,
+            EncryptionMethod::Aes192Ctr,
+            EncryptionMethod::Aes256Ctr,
+        ];
+        let mut plaintext = vec![0; 10240];
+        OsRng.fill_bytes(&mut plaintext);
+        let offsets = [1024, 1024 + 1, 10240 - 1];
+        let sizes = [1024, 8000];
+        let written = vec![0; 10240];
+        for method in methods {
+            let key = generate_data_key(method);
+            let writable_text = std::io::Cursor::new(written.clone());
+            let iv = Iv::new_ctr();
+            let encrypter = EncrypterWriter::new(writable_text, method, &key, iv).unwrap();
+            let mut decrypter = DecrypterWriter::new(encrypter, method, &key, iv).unwrap();
```

Comment:
Maybe we need to prove here that the previous encryption was successful

Reply:
You mean encrypted != plaintext?